### PR TITLE
Metadata as keymap

### PR DESCRIPTION
**Pull request by fgalan**
Implements #1112 

This PR include changes in the code and data model documentation. Migration script and related documentation will be part of a next PR.

@kzangeli

### CODE REVIEW

File: test/functionalTest/cases/1112_metadata_as_keymap/metadata_with_dots_v1.test [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+echo "02. Retieve entity"
+echo "=================="
+orionCurl --url "/v1/contextEntities/E"
+echo
+echo
+
+
+echo "03. Check the '=' in DB"
+echo "======================="
+mongoCmd ${CB_DB_NAME} 'db.entities.find({ "_id" : { "id" : "E",  "type": "T", "servicePath" : "/" } }, {_id:0, attrs: 1})'
```
Comment (fgalan) [inline test/functionalTest/cases/1112_metadata_as_keymap/metadata_with_dots_v1.test]: What is important to review here is that the key in the `md` maps correctly uses `=` as encoding character for dots.

File: test/functionalTest/cases/1112_metadata_as_keymap/metadata_with_dots_v1.test [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    "statusCode": {
+        "code": "200",
+        "reasonPhrase": "OK"
+    }
+}
+
+
+03. Check the '=' in DB
+=======================
+{ "attrs" : { "A" : { "type" : "T", "creDate" : REGEX(.*), "value" : "foo", "md" : { "m=x" : { "type" : "string", "value" : "black" }, "m=y" : { "type" : "string", "value" : "sweet" } }, "mdNames" : [ "m.x", "m.y" ] } } }
```
Comment (fgalan) [inline test/functionalTest/cases/1112_metadata_as_keymap/metadata_with_dots_v1.test]: What is important to review here is that the key in the `md` maps correctly uses `=` as encoding character for dots.

File: test/functionalTest/cases/1112_metadata_as_keymap/metadata_with_dots_v2.test [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        "value": "foo"
+    },
+    "id": "E",
+    "type": "T"
+}
+
+
+03. Check the '=' in DB
+=======================
+{ "attrs" : { "A" : { "type" : "T", "creDate" : REGEX(.*), "value" : "foo", "md" : { "m=x" : { "type" : "string", "value" : "black" }, "m=y" : { "type" : "double", "value" : 5 } }, "mdNames" : [ "m.x", "m.y" ] } } }
```
Comment (fgalan) [inline test/functionalTest/cases/1112_metadata_as_keymap/metadata_with_dots_v2.test]: What is important to review here is that the key in the `md` maps correctly uses `=` as encoding character for dots.

File: test/functionalTest/cases/2339_json_native_types_in_v1/json_native_types_in_v1.test [hunk 4]
```diff
@@ -274,29 +274,29 @@ Date: REGEX(.*)
                     {
                         "metadatas": [
                             {
```
Comment (fgalan) [inline test/functionalTest/cases/2339_json_native_types_in_v1/json_native_types_in_v1.test]: Note that the changes in this .test are only on the ordering of metadata (note the Content-Lenght hasn't changed).

File: doc/manuals/admin/database_model.md [hunk 5]
```diff
@@ -50,14 +50,21 @@ Fields:
...
         manual](../user/structured_attribute_valued.md)).
-    -   **md** (optional): custom metadata. This is a vector of metadata
-        objects, each one with a **name**, **type** and **value**.
+    -   **md** (optional): custom metadata. This is a keymap of metadata
+        objects. The key is generated with the metadata
+        name (changing "." for "=", as "." is not a valid character in
+        MongoDB document keys), e.g. the metadata with name "m.x" will
+        use the following key: "m=x". The object value per each key has two
+        fields: **type** and **value** of the metadata.
+    -   **mdNames*: an array of strings. Its elements are the names of the
```
Comment (kzangeli) [inline doc/manuals/admin/database_model.md]: Too much bold here, right?

File: src/lib/mongoBackend/MongoCommonUpdate.cpp [hunk 6]
```diff
@@ -241,46 +226,35 @@ static bool equalMetadataValues(BSONObj& md1, BSONObj& md2)
...
+* equalMetadata -
 *
-* Given two vectors with the same metadata names-types, check that all the values
+* Given two metadata object assuming they have the same size (i.e. number of elements). check that all the values
 * are equal, returning false otherwise.
 *
 * Arguments are passed by reference to avoid "heavy" copies
 */
-static bool equalMetadataVectors(BSONObj& mdV1, BSONObj& mdV2)
+static bool equalMetadata(BSONObj& md1, BSONObj& md2)
```
Comment (kzangeli) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: Could be a method in the Metadata class:

    if (md1.equals(md2))
    {
    }

Easier to find the function later if its in the class. Less prone for copies.
If you wish to keep all mongo stuff (BSONObj) encapsulated in mongBackend, a method AND a function would be necessary.

File: src/lib/mongoBackend/MongoCommonUpdate.cpp [hunk 7]
```diff
@@ -241,46 +226,35 @@ static bool equalMetadataValues(BSONObj& md1, BSONObj& md2)
...
+    BSONObj md2Item = getFieldF(md2, currentMd).embeddedObject();
+
+    if (!equalMetadataValues(md1Item, md2Item))
     {
-      /* Shortcut for early passing to the next attribute in i1 */
-      found = false;
-      continue;  // loop in i1
+      return false;
     }
   }
```
Comment (kzangeli) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: Where is the final 'return true' ?

File: src/lib/mongoBackend/MongoCommonUpdate.cpp [hunk 8]
```diff
@@ -365,32 +339,36 @@ bool attrValueChanges(BSONObj& attr, ContextAttribute* caP, std::string apiVersi
...
   if (!mdP->typeGiven && useDefaultType)
   {
     type = DEFAULT_TYPE;
   }
+
+  mdNamesBuilder->append(mdP->name);
+  std::string effectiveName = dbDotEncode(mdP->name);
+
   if (type != "")
   {
```
Comment (kzangeli) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: This empty line is not allowed by the style_check

File: src/lib/mongoBackend/MongoCommonUpdate.cpp [hunk 9]
```diff
@@ -504,37 +483,41 @@ static bool mergeAttrInfo(BSONObj& attr, ContextAttribute* caP, BSONObj* mergedA
...
   /* Second, for each metadata previously in the metadata vector but *not included in the request*, add it as is */
-  int      mdVSize = 0;
-  BSONObj  mdV;
-
+  int      mdSize = 0;
+  BSONObj  md;
   if (attr.hasField(ENT_ATTRS_MD))
-  {
-    mdV = getFieldF(attr, ENT_ATTRS_MD).embeddedObject();
+  {
```
Comment (kzangeli) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: spaces

File: test/functionalTest/cases/1112_metadata_as_keymap/metadata_with_dots_v1.test [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+Metadata with dot in the name (NGSIv1)
+
+--SHELL-INIT--
+dbInit CB
+brokerStart CB
+
+--SHELL--
+
+# 01. Create entity with metadata with dot in the name
+# 02. Retieve entity
```
Comment (kzangeli) [inline test/functionalTest/cases/1112_metadata_as_keymap/metadata_with_dots_v1.test]: typo: Retrieve

File: test/functionalTest/cases/1112_metadata_as_keymap/metadata_with_dots_v1.test [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+        "isPattern": "false",
+        "id": "E",
+        "attributes": [
+        {
+            "name": "A",
+            "type": "T",
+            "value": "foo",
+            "metadatas": [
+                {
+                    "name": "m.x",
```
Comment (kzangeli) [inline test/functionalTest/cases/1112_metadata_as_keymap/metadata_with_dots_v1.test]: Is '=' a forbidden char in metadata name?
I imagine it is. If not we would have chosen some other char, I hope :-)

File: test/functionalTest/cases/1112_metadata_as_keymap/metadata_with_dots_v2.test [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+Metadata with dot in the name (NGSIv2)
+
+--SHELL-INIT--
+dbInit CB
+brokerStart CB
+
+--SHELL--
+
+# 01. Create entity with metadata with dot in the name
+# 02. Retieve entity
```
Comment (kzangeli) [inline test/functionalTest/cases/1112_metadata_as_keymap/metadata_with_dots_v2.test]: Retrieve

File: doc/manuals/admin/database_model.md (hunk 5 above)
Comment (kzangeli) [inline doc/manuals/admin/database_model.md]: Fixed in 5c5a0e5

File: test/functionalTest/cases/1112_metadata_as_keymap/metadata_with_dots_v1.test (hunk 2 above)
Comment (fgalan) [inline test/functionalTest/cases/1112_metadata_as_keymap/metadata_with_dots_v1.test]: NTC (just a remark)

File: test/functionalTest/cases/1112_metadata_as_keymap/metadata_with_dots_v2.test (hunk 3 above)
Comment (fgalan) [inline test/functionalTest/cases/1112_metadata_as_keymap/metadata_with_dots_v2.test]: NTC (just a remark)

File: test/functionalTest/cases/2339_json_native_types_in_v1/json_native_types_in_v1.test (hunk 4 above)
Comment (fgalan) [inline test/functionalTest/cases/2339_json_native_types_in_v1/json_native_types_in_v1.test]: NTC (just a remark)

File: src/lib/mongoBackend/MongoCommonUpdate.cpp (hunk 7 above)
Comment (fgalan) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: At line 261, a little down below (you probably need to expand diff at github web to see it).

File: test/functionalTest/cases/1112_metadata_as_keymap/metadata_with_dots_v1.test (hunk 11 above)
Comment (fgalan) [inline test/functionalTest/cases/1112_metadata_as_keymap/metadata_with_dots_v1.test]: Yes, it is: https://fiware-orion.readthedocs.io/en/develop/user/forbidden_characters/index.html

File: src/lib/mongoBackend/MongoCommonUpdate.cpp (hunk 8 above)
Comment (fgalan) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: Fixed in d4fef947abb1695a87b687793597f5fb74485fac

File: src/lib/mongoBackend/MongoCommonUpdate.cpp (hunk 9 above)
Comment (fgalan) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: Fixed in d4fef947abb1695a87b687793597f5fb74485fac

File: test/functionalTest/cases/1112_metadata_as_keymap/metadata_with_dots_v1.test (hunk 10 above)
Comment (fgalan) [inline test/functionalTest/cases/1112_metadata_as_keymap/metadata_with_dots_v1.test]: Fixed in d4fef947abb1695a87b687793597f5fb74485fac

File: test/functionalTest/cases/1112_metadata_as_keymap/metadata_with_dots_v2.test (hunk 12 above)
Comment (fgalan) [inline test/functionalTest/cases/1112_metadata_as_keymap/metadata_with_dots_v2.test]: Fixed in d4fef947abb1695a87b687793597f5fb74485fac

File: src/lib/mongoBackend/MongoCommonUpdate.cpp (hunk 7 above)
Comment (kzangeli) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: ok, NTC

File: test/functionalTest/cases/1112_metadata_as_keymap/metadata_with_dots_v1.test (hunk 11 above)
Comment (kzangeli) [inline test/functionalTest/cases/1112_metadata_as_keymap/metadata_with_dots_v1.test]: NTC

**Comment by kzangeli** [comment]
LGTM

**Pull request merged**